File: solutions/102 - Binary Tree Level Order Traversal/BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <sstream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        // Time Complexity: O(N) because we need to process each node
        // Space Complexity: O(N) for the answer and queue
        vector<vector<int>> ans;
        if (root == nullptr)
        {
            return ans;
        }
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty())
        {
            int qSize = q.size();
            vector<int> vec;
            for (int i = 0; i < qSize; i++)
            {
                TreeNode* node = q.front();
                q.pop();
                vec.push_back(node->val);
                if (node->left != nullptr)
                {
                    q.push(node->left);
                }
                if (node->right != nullptr)
                {
                    q.push(node->right);
                }
            }
            ans.push_back(vec);
        }
        return ans;
    }
};

int main() {
    Solution sol;

    cout << "Enter the tree: ";
    string line;
    getline(cin, line);

    if (line.empty() || line == "[]") {
        cout << "Output:\n[]\n";
        return 0;
    }

    stringstream ss(line);
    vector<string> nodes;
    string tok;

    while (ss >> tok) nodes.push_back(tok);
    if (nodes.empty() || nodes[0] == "null") {
        cout << "Output:\n[]\n";
        return 0;
    }

    TreeNode* root = new TreeNode(stoi(nodes[0]));
    queue<TreeNode*> q;
    q.push(root);

    int i = 1;
    while (!q.empty() && i < nodes.size()) {
        TreeNode* cur = q.front();
        q.pop();

        if (i < nodes.size() && nodes[i] != "null") {
            cur->left = new TreeNode(stoi(nodes[i]));
            q.push(cur->left);
        }
        i++;

        if (i < nodes.size() && nodes[i] != "null") {
            cur->right = new TreeNode(stoi(nodes[i]));
            q.push(cur->right);
        }
        i++;
    }

    vector<vector<int>> res = sol.levelOrder(root);

    cout << "Output:\n";
    for (auto& level : res) {
        for (int v : level) cout << v << " ";
        cout << "\n";
    }

    return 0;
}

File: solutions/102 - Binary Tree Level Order Traversal/BFS.java
import java.util.*;
// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        // Time Complexity: O(N) because we visit each node once.
        // Space Complexity: O(N) for the answer and queue.
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            List<Integer> currLevel = new ArrayList<>();
            int queueSize = queue.size();
            for (int i = 0; i < queueSize; i++) {
                TreeNode currNode = queue.poll();
                currLevel.add(currNode.val);
                if (currNode.left != null) {
                    queue.add(currNode.left);
                }
                if (currNode.right != null) {
                    queue.add(currNode.right);
                }
            }
            ans.add(currLevel);
        }
        return ans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter binary tree: ");
        String line = sc.nextLine().trim();

        if (line.isEmpty() || line.equals("[]")) {
            System.out.println("Output:");
            System.out.println("[]");
            sc.close();
            return;
        }

        String[] values = line.split("\\s+");
        TreeNode root = null;

        if (!values[0].equalsIgnoreCase("null")) {
            root = new TreeNode(Integer.parseInt(values[0]));
            Queue<TreeNode> q = new ArrayDeque<>();
            q.add(root);

            int i = 1;
            while (!q.isEmpty() && i < values.length) {
                TreeNode cur = q.poll();

                if (i < values.length && !values[i].equalsIgnoreCase("null")) {
                    cur.left = new TreeNode(Integer.parseInt(values[i]));
                    q.add(cur.left);
                }
                i++;

                if (i < values.length && !values[i].equalsIgnoreCase("null")) {
                    cur.right = new TreeNode(Integer.parseInt(values[i]));
                    q.add(cur.right);
                }
                i++;
            }
        }

        Solution sol = new Solution();
        List<List<Integer>> res = sol.levelOrder(root);

        System.out.println("Output: ");
        for (List<Integer> level : res) {
            for (int v : level) {
                System.out.print(v + " ");
            }
            System.out.println();
        }

        sc.close();
    }
}